spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/QueryPushdownIntegrationTestBase.java

```diff
@@ -432,4 +405,340 @@ public void testAggregateExpressions() {
+        spark
+            .createDataset(
+                Arrays.asList(
+                    new NumStruct(
+                        1L,
+                        4L,
+                        3L,
+                        ImmutableList.of(new StringStruct("3:str3", "4:str4", "3:str3"))),
+                    new NumStruct(
+                        3L,
+                        5L,
+                        6L,
+                        ImmutableList.of(new StringStruct("6:str6", "5:str5", "3:str3"))),
+                    new NumStruct(
+                        3L,
+                        1L,
+                        4L,
+                        ImmutableList.of(new StringStruct("3:str3", "1:str1", "4:str4")))),
+                Encoders.bean(NumStruct.class))
+            .toDF();
+    String tableForJoin = testTable + "_to_join";
+    df_to_join
+        .write()
+        .format("bigquery")
+        .mode(SaveMode.Append)
+        .option("table", testDataset.toString() + "." + tableForJoin)
+        .option("temporaryGcsBucket", TestConstants.TEMPORARY_GCS_BUCKET)
+        .option("writeMethod", WriteMethod.INDIRECT.toString())
+        .save();
+
+    df_to_join =
+        spark
+            .read()
+            .format("bigquery")
+            .option("materializationDataset", testDataset.toString())
+            .load(testDataset.toString() + "." + tableForJoin);
+
+    df_to_join.createOrReplaceTempView("numStructDF_to_join");
+
+    List<Row> result;
+    result = df.join(df_to_join, df.col("num1").equalTo(df_to_join.col("num1"))).collectAsList();
+    /*
+     +----+----+----+--------------------+----+----+----+--------------------+
+     |num1|num2|num3|             strings|num1|num2|num3|             strings|
+     +----+----+----+--------------------+----+----+----+--------------------+
+     |   4|   3|   2|[[[[3:str1, 4:str...|   4|   1|   3|[[[[1:str1, 4:str...|
+     |   3|   2|   1|[[[[2:str1, 3:str...|   3|   4|   3|[[[[4:str4, 3:str...|
+     +----+----+----+--------------------+----+----+----+--------------------+
+    */
+    assertThat(result.size()).isEqualTo(2);
+    for (Row r : result) {
+      assertThat(r.get(0)).isEqualTo(r.get(4));
+    }
+    result = df_to_join.join(df, df.col("num1").equalTo(df_to_join.col("num1"))).collectAsList();
+    /*
+     +----+----+----+--------------------+----+----+----+--------------------+
+     |num1|num2|num3|             strings|num1|num2|num3|             strings|
+     +----+----+----+--------------------+----+----+----+--------------------+
+     |   4|   1|   3|[[[[1:str1, 4:str...|   4|   3|   2|[[[[3:str1, 4:str...|
+     |   3|   4|   3|[[[[4:str4, 3:str...|   3|   2|   1|[[[[2:str1, 3:str...|
+     +----+----+----+--------------------+----+----+----+--------------------+
+    */
+    assertThat(result.size()).isEqualTo(2);
+    for (Row r : result) {
+      assertThat(r.size()).isEqualTo(8);
+      assertThat(r.get(0)).isEqualTo(r.get(4));
+    }
+
+    result =
```

Comment:
Can we have separate methods for each of the different types of joins? Would be easier to debug

Reply:
Sure